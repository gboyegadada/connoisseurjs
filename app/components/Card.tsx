import React, { ReactNode, FC } from 'react'
import { ResultItem } from '../types/search'
import { MdLocationCity, MdLocationOn } from 'react-icons/md'

interface CardProps {
    children?: ReactNode
    data: ResultItem,
    key: string
}

const Card: FC<CardProps> = (props) => {
    const { uniqueId, title, uri, raw } = props.data
    const numFormat = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'CAD' });

    return (
        <li className='card' key={uniqueId}>
            <a href={uri}>
                <img className='card-image' src={raw.tpthumbnailuri} />
            </a>
            <a href={uri}>
                <h5>{title}</h5>
            </a>
            <ul className='card-info'>
                <li><MdLocationOn /> {raw.tppays}</li>
                <li><MdLocationCity /> {raw.tpregion}</li>
                <li>
                { numFormat.format(parseFloat(raw.tpprixnum)) }
                </li>
            </ul>
            <ul className='tags'></ul>
        </li>
    )
}

export default Card