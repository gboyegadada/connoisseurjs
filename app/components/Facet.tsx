import React, { ReactNode, FC } from 'react'
import { FacetItem } from '../types/facet'

interface FacetProps {
    children?: ReactNode
    data: FacetItem,
    key: string
}

const Facet: FC<FacetProps> = (props) => {
    const { field } = props.data
    const numFormat = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'CAD' });

    return (
        <li className='facet' key={field} data-field={field}>
            <h5>&nbsp;</h5>
        </li>
    )
}

export default Facet