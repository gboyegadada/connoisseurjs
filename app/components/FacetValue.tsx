import React, { ReactNode, FC } from 'react'
import { FacetItem, FacetValue } from '../types/facet'
import { MdCheckBox, MdCheckBoxOutlineBlank } from 'react-icons/md';
import { createFacetConnect } from '../redux/store';

interface FacetValueProps {
    children?: ReactNode
    data: FacetValue,
    key: string
    field: string
}

const FacetConnect = createFacetConnect({
    mapActions: actions => ({
        tick: actions.tickFacetValue,
        untick: actions.untickFacetValue
    })
})

const FacetValue: FC<FacetValueProps> = (props) => {
    const { key, field, data: { value, numberOfResults, checked, exclude } } = props
    const numFormat = new Intl.NumberFormat('en-IN');

    const toggleTick = (a: any) => {
        checked ? a.untick({ field, value }) : a.tick({ field, value })
    }

    return (
        <FacetConnect>
        {(_, actions) => (
        <li className='facet-value' key={key} data-value={value} data-count={numberOfResults} onClick={() => toggleTick(actions)}>
            { checked ? <MdCheckBox size={21} /> : <MdCheckBoxOutlineBlank size={21} /> }
            
            <span className="facet-value-label">{value}</span> <span className="facet-value-count">{numFormat.format(numberOfResults)}</span>
        </li>
        )}
        </FacetConnect>
    )
}

export default FacetValue