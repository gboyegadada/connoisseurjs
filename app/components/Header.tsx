import React, { ReactNode, FC, useState } from 'react'
import { createSearchConnect, createMenuConnect } from '../redux/store'
import { SearchStatus } from '../types/search'
import { MdMenu } from 'react-icons/md'
import SearchBox from './SearchBox'


interface HeaderProps {
    children?: ReactNode
    className?: string
}

const SearchConnect = createSearchConnect({
    mapActions: actions => ({
        startSearch: actions.startSearch
    }),
    mapState: selectors => ({
        status: selectors.getStatus()
    })
})

const MenuConnect = createMenuConnect({
    mapActions: actions => ({
        toggleMenu: actions.toggleMenu
    }),
    mapState: menuOpenState => ({
        menuOpen: menuOpenState
    })
})

const Header: FC<HeaderProps> = () => {
    return (
        <SearchConnect>
            {(_, actions) => (
                <header>
                    <MenuConnect >
                    {(_m, menuActions) => (
                        <button className='menu-toggle' onClick={() => menuActions.toggleMenu()}>
                            <MdMenu size={24} />
                        </button>
                    )}
                    </MenuConnect>

                    <SearchBox onSubmit={actions.startSearch} status={_.status} />
                </header>
            )}
        </SearchConnect>
    )
}

export default Header