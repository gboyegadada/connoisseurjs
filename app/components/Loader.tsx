import React, { ReactNode, FC } from 'react'

interface LoaderProps {
    children?: ReactNode
    className?: string
    loading: boolean
    error?: boolean 
}

const Loader: FC<LoaderProps> = () => {
    return (
        <h3>Loading...</h3>
    )
}

export default Loader