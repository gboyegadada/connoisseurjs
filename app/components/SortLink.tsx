import React, { FC, useState } from 'react'
import { Link } from 'react-router-dom'
import queryString from 'query-string'
import { MdExpandMore, MdExpandLess } from 'react-icons/md'

interface SortLinkProps {
    label: string
    field?: string
    criteria?: string
    activeField: string
    activeCriteria: string
    queryParams: {
        q: string
        aq: string
    }
    sortAction: Function
}

const SortLink: FC<SortLinkProps> = ({ label, criteria, field, activeField, activeCriteria, sortAction }) => {
    // Grab @field and @order from current query string if any...
    const q = queryString.parse(location.search)
    let {order = 'ascending'} = q

    if (!field) {
        order = criteria
    }
    else if (order !== 'ascending' && order !== 'descending') {
        order = 'ascending'
    } 

    // Inititial state: ascending || descending || relevancy
    const [_order, setOrder] = useState(order)

    // Setup query params for this button
    // Something like this: "&sort=@field%20ascending"
    const s = {
        sort: field ? field : '',
        order: toggleOrder(_order)
    }

    // Determine if button is active
    const active = (field && field === activeField) || (!field && criteria && criteria === activeCriteria) 

    function toggleOrder(o: any) {
        switch (o) {
            case 'ascending':
                return 'descending'
                break
            case 'descending':
                return 'ascending'
                break
            default:
                return o
                break
        }
    }

    function handleClick() {

        // Toggle sort order
        const o = toggleOrder(_order)
        setOrder(o)

        // Dispatch sort action
        let payload: {field?: string, criteria?: string } = {}
        
        if (field) {
            payload.field = field
            payload.criteria = `field${o}`
        } else {
            payload.criteria = o
        }

        sortAction(payload)
    }

    return (
        <li>            
        <Link 
            className={active ? 'active' : ''} 
            to={`/?${queryString.stringify({ ...q, ...s })}`}
            onClick={handleClick}>
                {label} 
                {active && _order === 'ascending' && <MdExpandMore size='1.3em' />}
                {active && _order === 'descending' && <MdExpandLess size='1.3em' />}
        </Link>
        </li>
    )
}

export default SortLink