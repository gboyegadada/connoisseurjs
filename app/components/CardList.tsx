import React, { ReactNode, FC, useState } from 'react'
import Card from './Card'
import { ResultItem } from '../types/search'
import ReactPaginate from 'react-paginate'
import { createPaginatorConnect } from '../redux/store'
import { MdInfoOutline, MdSwapVert } from 'react-icons/md'

interface CardListProps {
    children?: ReactNode
    list: ResultItem[]
}

const PaginatorConnect = createPaginatorConnect({
    mapActions: actions => ({
        next: actions.next,
        previous: actions.previous,
        gotoPage: actions.gotoPage
    }),
    mapState: selectors => ({
        offset: selectors.getOffset(),
        pageLength: selectors.getPageLength(),
        pageCount: selectors.getPageCount(),
        totalCount: selectors.getTotalCount(),
        itemCount: selectors.getItemCount(),
        q: selectors.getSearchQuery(),
        queryParams: selectors.getQueryParams()
    })
})

const CardList: FC<CardListProps> = (props) => {
    const { list } = props

    const [pageCount, setPageCount] = useState(0)
    const numFormat = new Intl.NumberFormat('en-US');

    return (
        <PaginatorConnect>
            {(_, actions) =>(
            <>
                <div className='top'>
                    <ul className='list-info'>
                        <li>
                            <MdInfoOutline size='1.2em' /> Items <strong>{(_.offset+1)}</strong> to <strong>{(_.offset + _.itemCount)}</strong> of <strong>{numFormat.format(_.totalCount)}</strong>
                        </li>

                        <li>
                            <MdSwapVert size='1.2em' />
                        </li>
                    </ul>
                </div>
                
                <ul className='grid card-list'>
                    {list.map(item => (
                        <Card key={item.uniqueId} data={item} />
                    ))}
                </ul>
            
                <div className='bottom flex-center'>
                <ReactPaginate
                    previousLabel={'previous'}
                    nextLabel={'next'}
                    breakLabel={'...'}
                    breakClassName={'break-me'}
                    pageCount={_.pageCount}
                    marginPagesDisplayed={2}
                    pageRangeDisplayed={5}
                    onPageChange={pg => actions.gotoPage({ page: pg.selected, query: {...{q: _.q}, ..._.queryParams} })}
                    containerClassName={'pagination'}
                    activeClassName={'active'}
                />
                </div>
            </>
            )}
        </PaginatorConnect>
    )
}

export default CardList