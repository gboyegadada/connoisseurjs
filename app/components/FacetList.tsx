import React, { ReactNode, FC } from 'react'
import Facet from './Facet'
import { FacetItem, FacetList } from '../types/facet'

interface FacetListProps {
    children?: ReactNode
    list: FacetList
}

const FacetList: FC<FacetListProps> = (props) => {
    const { list } = props

    return (
        <ul className='grid facet-list'>
            { Object.keys(list).map(id => <Facet key={list[id].field} data={list[id]} />) }
        </ul>
    )
}

export default FacetList