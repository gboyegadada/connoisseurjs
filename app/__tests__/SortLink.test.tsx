import React from 'react';
import { mount } from 'enzyme'

import { SortLink } from "../components/SortLink";

jest.mock('react-router-dom');

describe("<SortLink />", () => {
  test("should display clickable sort links", async () => {
    const wrapper = mount(<SortLink
          label='Millesime' 
          field='@tpmillesime' 
          criteria='ascending'
          activeField='@tpmillesime'
          activeCriteria='fieldascending'
          queryParams={{
              q: '',
              aq: '',
          }}
          sortAction={(): any => null}
        />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('Link').length).toEqual(1)
  })
});