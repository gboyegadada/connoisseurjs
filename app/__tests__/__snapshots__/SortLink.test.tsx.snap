// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SortLink /> should display clickable sort links 1`] = `
<Component
  activeCriteria="fieldascending"
  activeField="@tpmillesime"
  criteria="ascending"
  field="@tpmillesime"
  label="Millesime"
  queryParams={
    Object {
      "aq": "",
      "q": "",
    }
  }
  sortAction={[Function]}
>
  <li>
    <Link
      className="active"
      onClick={[Function]}
      to="/?sort=%40tpmillesime&order=descending"
    />
  </li>
</Component>
`;
